lib/IDE/Formatting.cpp

```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+            return Stop;
+        } else {
+          if (!handleParens(TE->getLParenLoc(), TE->getRParenLoc(), ContextLoc))
+            return Stop;
+        }
+        if (TE->hasTrailingClosure()) {
+          if (auto *Last = TE->getElements().back()) {
+            TC = dyn_cast<ClosureExpr>(Last);
+            TCL = dyn_cast<CaptureListExpr>(Last);
+          }
+        }
+      }
+      if (TC && !handleBraceStmt(TC->getBody(), ContextLoc))
+        return Stop;
+      if (TCL && !handleBraceStmt(TCL->getClosureBody()->getBody(), ContextLoc))
+        return Stop;
+    }
+    return Continue;
+  }
+
+  std::pair<bool, Pattern*> walkToPatternPre(Pattern *P) override {
+    std::pair<bool, Pattern*> Continue = {true, P}, Stop = {false, nullptr};
+
+    if (P->isImplicit())
+      return Continue;
+
+    if (isa<TuplePattern>(P) || isa<ParenPattern>(P)) {
+      if (!handleParens(P->getStartLoc(), P->getEndLoc(), P->getStartLoc()))
+        return Stop;
+    }
+
+    return Continue;
+  }
+
+  bool walkToTypeReprPre(TypeRepr *T) override {
+    bool Continue = true, Stop = false;
+
+    if (auto *TT = dyn_cast<TupleTypeRepr>(T)) {
+      SourceRange Parens = TT->getParens();
+      if (!handleParens(Parens.Start, Parens.End, Parens.Start))
+        return Stop;
+    } else if (isa<ArrayTypeRepr>(T) || isa<DictionaryTypeRepr>(T)) {
+      if (!handleSquares(T->getStartLoc(), T->getEndLoc(), T->getStartLoc()))
+        return Stop;
+    } else if (auto *GI = dyn_cast<GenericIdentTypeRepr>(T)) {
+      SourceLoc ContextLoc = GI->getNameLoc().getBaseNameLoc();
+      SourceRange Brackets = GI->getAngleBrackets();
+      if (!handleAngles(Brackets.Start, Brackets.End, ContextLoc))
+        return Stop;
+    }
+
+    return Continue;
+  }
+
+  bool shouldWalkIntoGenericParams() override { return true; }
+};
+
+/// Indicates whether a range is an open or closed range.
+enum class RangeKind {Closed, Open};
+
+/// A helper class that determines whether a given node, or subrage of a node
+/// should indent or not when it spans multiple lines.
+class OutdentChecker: protected RangeWalker {
+  SourceRange CheckRange; ///< The source range to consider.
+  RangeKind CheckRangeKind; ///< Whether \c CheckRange is open or closed.
+  bool IsOutdenting = false; ///< Tracks whether a seen range prevents indenting.
+  llvm::DenseMap<SourceLoc, ContextOverride> LineStartToOverride;
+
+  explicit OutdentChecker(SourceManager &SM,
+                          SourceRange CheckRange,  RangeKind CheckRangeKind)
+  : RangeWalker(SM), CheckRange(CheckRange), CheckRangeKind(CheckRangeKind) {
+    assert(CheckRange.isValid());
+  }
+
+  bool handleRange(SourceLoc L, SourceLoc R, SourceLoc ContextLoc) override {
+    assert(L.isValid() && ContextLoc.isValid());
+
+    // Ignore parens/braces/brackets outside of the open/closed check range.
+    if (!isInCheckRange(L, R))
+      return true;
+
+    // The CheckRange is made outdenting by any parens/braces/brackets with:
```

Comment:
Thanks, this is still complicated, but the examples help a lot to get a feeling for how it works!